### PR TITLE
feat(i18n): multi-lingual texts for config-like entities

### LINKED ISSUE #3862: i18n: multi-lingual config (configure labels in multiple languages)
_As an admin I want to configure my system in more than one language in order to give my team members the app in their different local languages._

Make the platform truly multilingual, so that within one system different users can use different languages at the same time.

**Context**
We already support translation of all static texts of the app, this includes buttons and fixed features. However, the majority of the UI is actually custom configuration by admin users (e.g. entity type names, field labels, etc.). At the moment, all these texts can only be configured once, so there is no option to show field labels in a different language to different users of the same system.

#### Displaying multi-lingual config:
- extend the config format so that as an alternative to simple `string`, a map `{ "language-code": "text", "en": "Welcome" }` is stored.
   - keep supporting simply one string to avoid migrations or unnecessarily complex config
- In the ConfigService, when loading the config doc, dynamically parse this mutli-lingual config.
   - To all other code, only expose the existing simple format, where the texts are the ones for the current selected language
- some additional entity types need the same kind of parsing of mulit-lingual config. Probably we should implement a service that handles this and can be reused:
   - `PublicFormConfig`
   - `ConfigurableEnum`
   - maybe `TemplateExport`
   - `EmailTemplate`
   - `"Config:SetupWizard"`

#### Configuring multi-lingual config
- any config form field that hold translatable text displays a small [icon](https://fontawesome.com/icons/language?f=classic&s=solid) that opens a dialog to configure all translations for that text (similar to configuring enum options)
   - the form field itself can still be used as a simple text field to configure the text for the currently active language
   - the components store a multilingual map instead of simple string, if active
- (?) as an additional option, an Admin view lists all translatable texts in a global table, where columns for each language give a quick way to review and update translations without having to open an endless number of field dialogs

Image

---
## Implementation Steps
1. Start with manually defining a multi-lingual config file, only implement the Admin UI later, after the format has been tested
2. Figure out a way to detect which config parts can/are multi-lingual (need to be parsed) and which are static (keys, IDs but also if the config is only using one lanuage)
3. ...

### PR DESCRIPTION
**Pull request by sadaf895**
- closes  #3862

## PR Checklist

_Before you request a manual PR review from someone, please go through all of this list:_

- [ ] manually tested (all) required functionality yourself and added comment with clear steps for manual testing
- [ ] reviewed the "Files changed" section of the PR briefly
- [ ] added label **"Final Review"** to trigger UI regression testing (Argos visual review)
- [ ] triggered CodeRabbit AI review by commenting **"@coderabbitai review"** (e2e tests run automatically on every push)
- [ ] marked each code review comment as resolved (or commented on it with a question, if unsure)

--> then mark the PR "Ready for Review"

---
## Manual Testing

1. - [ ] Open any record with a dropdown (e.g. a Child's **Gender**, or a Note's **Category**)
- [x] Click the wrench icon on the dropdown to open **Edit dropdown options**
- [x] Click language icon next to an option's label → enter a German text → Save → Save the dialog
- [x] Reopen language dialog on that option → **both languages are still there**
- [x] Now change something unrelated on the same dropdown (e.g. an option's **colour**) and save
- [x] Reopen dialog again → **both languages are still there**

If German disappears at either point, the save flattened it — that's a blocker.

2. Dropdown options appear in the user's language

- [x] Switch the language to **Deutsch**
- [x] The option you translated shows its **German** label
- [x] Options you did *not* translate still show English (fallback works)
- [x] Switch back to English → everything reverts

3. Reports

- [x] Admin → **Reports** → open a report → edit **Title** → language dialog → add a German text → save
- [x] Reports page → the report dropdown shows the German title
- [ ] Run the report → **Export** → the file name uses the German title, not `[object Object]`

4. Progress widget

- [x] Dashboard → progress widget →  (wrench)
- [x] The **title** and each **task label** have a language icon button
- [x] Add a German text, save, and confirm the widget shows it in German

5. The dashboard editor must not destroy translations

- [x] In **English**, go to Admin → **Edit Dashboard Widgets**
- [x] Just **drag a widget to reorder it** (change nothing else) → **Save**
- [x] Switch the language to **Deutsch**
- [x] The dashboard shortcut labels are still German.
  If they show English, the editor flattened the translations on save — that's a blocker.

6.  Admin tabs

- [x] Admin → **Configure Data Structure** → **Child** → **Details View**
- [x] Every tab shows a readable name — **no `[object Object]`** on the inactive tabs
- [x] Click through the tabs; each stays readable when selected and unselected

- [ ] 7. Email Template

### CODE REVIEW

**Comment by george-neha** [comment]
1. If translation is added to the dropdown values, can it be highlighted to maintain consistency across the system
image

**Comment by george-neha** [comment]
Sometimes on clicking the colour gradient pin wheel pops up. The click has to be very specific to open the translate option: 
Screenshot 2026-09-04 at 11 59 35 AM

**Comment by sadaf895** [comment]
@coderabbitai review

**Comment by sadaf895** [comment]
@coderabbitai review

**Review by Abhinegi2 — COMMENTED** [review]
Review notes on the blocking items only. Everything else (edge cases, manual-test gaps) kept out of the diff on purpose.

File: src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts [hunk 1]
```diff
@@ -28,11 +29,23 @@ export class ConfigurableEnumService {
...
   getEnumValues<T extends ConfigurableEnumValue = ConfigurableEnumValue>(
     id: string,
   ): T[] {
     const configurableEnum = this.getEnum(id);
-    return configurableEnum ? (configurableEnum.values as T[]) : [];
+    if (!configurableEnum) {
+      return [];
+    }
+
+    return configurableEnum.values.map((option) => ({
```
Comment (Abhinegi2) [inline src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts:45]: `getEnumValues()` now rebuilds the whole option list on every call, and it sits on a per-record hot path: `ConfigurableEnumDatatype.transformToObjectFormat` calls it once per enum field value while loading entities. A list of 2,000 records with two enum fields backed by a 30-option enum allocates ~120,000 objects, where before it reused the cached array.

Could we cache the resolved array per enum id (build it in `cacheEnum()`, clear it on update) and return that? Then the resolution cost is paid once per enum instead of once per field value.

File: src/app/core/entity-list/entity-list/entity-list.component.html [hunk 2]
```diff
@@ -233,15 +233,17 @@ <h2>{{ title() }}</h2>
     <button
       mat-menu-item
       (click)="copyPublicFormLinkForEntityType(formConfig)"
-      [matTooltip]="formConfig.title"
+      [matTooltip]="publicFormTitle(formConfig)"
```
Comment (Abhinegi2) [inline src/app/core/entity-list/entity-list/entity-list.component.html:236]: `publicFormTitle()` is a method call in a binding, so it re-runs on every change-detection cycle (and again on line 245), each time walking `Object.entries()` inside `resolveTranslatableText`. This was a plain property read before.

Resolving the titles once when `publicFormConfigs` is loaded, or exposing a `computed` with the resolved titles, would avoid that.

File: src/app/core/entity-list/entity-list/entity-list.component.html [hunk 3]
```diff
@@ -233,15 +233,17 @@ <h2>{{ title() }}</h2>
...
       i18n-matTooltip
       matTooltipPosition="before"
     >
       <fa-icon
         class="color-accent standard-icon-with-text"
         icon="link"
       ></fa-icon>
-      <span i18n>Copy Public Form Link</span> ({{ formConfig.title }})
+      <span i18n>Copy Public Form Link</span> ({{
+        publicFormTitle(formConfig)
```
Comment (Abhinegi2) [inline src/app/core/entity-list/entity-list/entity-list.component.html:245]: Second call site of the same method binding - same fix as above.

File: src/app/features/public-form/custom-form-link-button/custom-form-link-button.component.html [hunk 4]
```diff
@@ -7,7 +7,7 @@
     mat-flat-button
     (click)="copyLink(matchingForm)"
     matTooltip="This link to the `{{
-      matchingForm?.title
+      formTitle(matchingForm)
```
Comment (Abhinegi2) [inline src/app/features/public-form/custom-form-link-button/custom-form-link-button.component.html:10]: Same as in the entity list: `formTitle()` in a binding re-runs on every change-detection cycle (twice, counting line 24). `matchingForm` is derived from signals, so a `computed` holding the resolved title works here.

File: src/app/features/public-form/custom-form-link-button/custom-form-link-button.component.html [hunk 5]
```diff
@@ -20,7 +20,9 @@
       class="standard-icon-with-text"
     ></app-fa-dynamic-icon>
     <span i18n
-      >Copy link to individualized public form ({{ matchingForm?.title }})</span
+      >Copy link to individualized public form ({{
+        formTitle(matchingForm)
```
Comment (Abhinegi2) [inline src/app/features/public-form/custom-form-link-button/custom-form-link-button.component.html:24]: Second call site of the same method binding - same fix as above.

File: src/app/core/admin/building-blocks/admin-tabs/admin-tabs.component.html [hunk 6]
```diff
@@ -28,7 +28,7 @@
               ></app-admin-section-header>
             } @else {
               <!-- only current tab can be renamed -->
-              {{ tab[tabTitleProperty()] }}
+              {{ tabTitle(tab) }}
```
Comment (Abhinegi2) [inline src/app/core/admin/building-blocks/admin-tabs/admin-tabs.component.html:31]: `tabTitle(tab)` re-runs on every change-detection cycle for every non-selected tab. `tabs` is a `model()`, so a `computed` mapping the tabs to their resolved titles (read by `tabIndex`) would avoid the repeated resolve.

File: src/app/features/public-form/public-form.component.html [hunk 7]
```diff
@@ -25,10 +25,18 @@ <h1 i18n>No access to public form "{{ formConfig()?.title }}"</h1>
     }
   </div>
 } @else {
+  @if (showLanguageSelect) {
+    <div class="language-select-container">
+      <app-language-select
```
Comment (Abhinegi2) [inline src/app/features/public-form/public-form.component.html:30]: `LanguageSelectComponent.changeLocale()` calls `LanguageService.switchLocale()`, which writes localStorage and then does `window.location.reload()`. On a public form that means a visitor who has already filled in part of the form loses everything they typed, with no warning.

Could we confirm before switching when the form is dirty, or restore the entered values after the reload?

File: src/app/core/admin/building-blocks/admin-tabs/admin-tabs.component.html (hunk 6 above)
Comment (sadaf895) [reply src/app/core/admin/building-blocks/admin-tabs/admin-tabs.component.html:31]: Same as on phase-1 PR https://github.com/Aam-Digital/ndb-core/pull/4283 a computed over tabs() doesn't work here.

File: src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts (hunk 1 above)
Comment (sadaf895) [reply src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts]: I made the validity check self-validating instead of clearing on update — the update event arrives late over the async changes feed, and it always carries a fresh entity anyway, so an explicit clear would be both too late and redundant

**Comment by Abhinegi2** [comment]
@sadaf895, could you check failing unit test cases?

File: src/app/core/admin/setup-wizard/setup-wizard-config.ts [hunk 8]
```diff
@@ -12,6 +12,7 @@ export interface SetupWizardConfig {
   steps: SetupWizardStep[];
 }
 
+/** stored, the texts may be per-language maps - resolved for display (#3862) */
```
Comment (Abhinegi2) [inline src/app/core/admin/setup-wizard/setup-wizard-config.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
/** stored, the texts may be per-language maps - resolved for display */
```

File: src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts [hunk 9]
```diff
@@ -28,11 +42,36 @@ export class ConfigurableEnumService {
     return this.enums.set(entity.getId(), entity);
   }
 
+  /**
+   * The options with their labels resolved to the active language, as copies so
+   * the cached entity keeps every language (#3862). Ids are never translated.
```
Comment (Abhinegi2) [inline src/app/core/basic-datatypes/configurable-enum/configurable-enum.service.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
   * the cached entity keeps every language. Ids are never translated.
```

File: src/app/core/basic-datatypes/configurable-enum/configurable-enum.ts [hunk 10]
```diff
@@ -39,13 +44,9 @@ export class ConfigurableEnum extends Entity {
...
     }
 
-    // check for duplicates
-    if (
-      this.values.some(
-        (v) =>
-          v.label.trim().toLowerCase() === option.label.trim().toLowerCase(),
-      )
-    ) {
+    // check for duplicates by the text of the active language (#3862)
```
Comment (Abhinegi2) [inline src/app/core/basic-datatypes/configurable-enum/configurable-enum.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
    // check for duplicates by the text of the active language
```

File: src/app/core/config/display-translatable-text/display-translatable-text.component.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { ChangeDetectionStrategy, Component, computed } from "@angular/core";
+import { ViewDirective } from "#src/app/core/entity/default-datatype/view.directive";
+import { DynamicComponent } from "../dynamic-components/dynamic-component.decorator";
+import { TranslatableText } from "../multi-lingual-config";
+import { resolveActiveText } from "../../language/active-locale";
+
+/** Display a configurable text in the user's active language (#3862). */
```
Comment (Abhinegi2) [inline src/app/core/config/display-translatable-text/display-translatable-text.component.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
/** Display a configurable text in the user's active language. */
```

File: src/app/core/config/translatable-text.datatype.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+import {
+  DefaultDatatype,
+  ExportColumnMapping,
+} from "../entity/default-datatype/default.datatype";
+import { EntitySchemaField } from "../entity/schema/entity-schema-field";
+import { resolveActiveText } from "../language/active-locale";
+import { TranslatableText } from "./multi-lingual-config";
+
+/**
+ * A user-facing text that admins can configure in several languages (#3862).
```
Comment (Abhinegi2) [inline src/app/core/config/translatable-text.datatype.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
 * A user-facing text that admins can configure in several languages.
```

File: src/app/core/dashboard/admin-dashboard/admin-dashboard.component.ts [hunk 13]
```diff
@@ -70,6 +67,8 @@ export class AdminDashboardComponent {
     () => PREFIX_VIEW_CONFIG + this.dashboardViewId(),
   );
 
+  // raw, because save() writes this back to the config document - a resolved
+  // value would drop every other configured language (#3862)
```
Comment (Abhinegi2) [inline src/app/core/dashboard/admin-dashboard/admin-dashboard.component.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
  // value would drop every other configured language
```

File: src/app/core/entity/model/entity.ts [hunk 14]
```diff
@@ -357,6 +358,9 @@ export class Entity {
         if (value?.label) {
           value = value.label;
         }
+        // a configured text may be a per-language map (#3862); anything else
```
Comment (Abhinegi2) [inline src/app/core/entity/model/entity.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
        // a configured text may be a per-language map; anything else
```

File: src/app/core/language/available-locales.ts [hunk 15]
```diff
@@ -0,0 +1,15 @@
+/**
+ * The locales the app can be used in.
+ *
+ * Separate from `languages.ts` because that builds a `ConfigurableEnum`, which
+ * extends `Entity` - and `Entity` needs the locale ids itself (see #3862).
```
Comment (Abhinegi2) [inline src/app/core/language/available-locales.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
 * extends `Entity` - and `Entity` needs the locale ids itself.
```

File: src/app/features/dashboard-widgets/progress-dashboard-widget/progress-dashboard/progress-dashboard-config.ts [hunk 16]
```diff
@@ -18,12 +18,14 @@
...
 import { DatabaseEntity } from "../../../../core/entity/database-entity.decorator";
 import { DatabaseField } from "../../../../core/entity/database-field.decorator";
+import { TranslatableText } from "../../../../core/config/multi-lingual-config";
 
 @DatabaseEntity("ProgressDashboardConfig")
 export class ProgressDashboardConfig extends Entity {
   static override isInternalEntity = true;
 
-  @DatabaseField() title: string = $localize`Progress Widget`;
+  /** may hold a per-language map, resolve for display only (#3862) */
```
Comment (Abhinegi2) [inline src/app/features/dashboard-widgets/progress-dashboard-widget/progress-dashboard/progress-dashboard-config.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
  /** may hold a per-language map, resolve for display only */
```

File: src/app/features/dashboard-widgets/progress-dashboard-widget/progress-dashboard/progress-dashboard.component.ts [hunk 17]
```diff
@@ -61,6 +62,15 @@ export class ProgressDashboardComponent {
 
   overallPercentage = computed(() => this.getOverallProgressPercentage());
 
+  // display only - `data` keeps the raw values because it is what gets saved (#3862)
```
Comment (Abhinegi2) [inline src/app/features/dashboard-widgets/progress-dashboard-widget/progress-dashboard/progress-dashboard.component.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
  // display only - `data` keeps the raw values because it is what gets saved
```

File: src/app/features/markdown-page/markdown-content.ts [hunk 18]
```diff
@@ -7,8 +8,6 @@ import { Entity } from "app/core/entity/model/entity";
  */
 @DatabaseEntity("MarkdownContent")
 export class MarkdownContent extends Entity {
-  /**
-   * The content with markdown formatting.
-   */
-  @DatabaseField() content: string;
+  /** markdown content, may hold a per-language map (#3862) */
```
Comment (Abhinegi2) [inline src/app/features/markdown-page/markdown-content.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
  /** markdown content, may hold a per-language map */
```

File: src/app/features/reporting/report-config.ts [hunk 19]
```diff
@@ -22,12 +24,13 @@ class ReportConfig extends Entity {
   static override readonly route = "admin/report-config";
   static override readonly icon: IconName = "chart-line";
 
-  /** human-readable title of the report */
+  /** human-readable title of the report (may be a per-language map, #3862) */
```
Comment (Abhinegi2) [inline src/app/features/reporting/report-config.ts]: Drop the issue number — comments here shouldn't reference ticket numbers.

```suggestion
  /** human-readable title of the report (may be a per-language map) */
```
